libpromises/evalfunction.c

```diff
@@ -6659,6 +6659,189 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
+                fp->name, line_number);
+            continue;
+        }
+
+        if (SeqLength(list) == 1 &&
+            strlen(SeqAt(list, 0)) == 0)
+        {
+            Log(LOG_LEVEL_DEBUG,
+                "%s: Found empty line at line %zu",
+                fp->name, line_number);
+            SeqDestroy(list);
+            continue;
+        }
+
+        if (num_columns == 0)
+        {
+            num_columns = SeqLength(list);
+
+            if (class_index >= num_columns)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "%s: Class expression index is out of bounds. Row "
+                    " length %zu, index %zu",
+                    fp->name, num_columns, class_index);
+                FnFailure();
+            }
+        }
+        else if (num_columns != SeqLength(list))
+        {
+            Log(LOG_LEVEL_WARNING,
+                "%s: Line %zu has incorrect amount of elements, "
+                "%zu instead of %zu. Line ignored.",
+                fp->name, line_number, SeqLength(list), num_columns);
+            SeqDestroy(list);
+            continue;
+        }
+
+        // First parsed line is set to be heading if has_heading is true
+        if (has_heading && heading == NULL)
+        {
+            Log(LOG_LEVEL_DEBUG, "%s: Found header at line %zu",
+                fp->name, line_number);
+            heading = list;
+            SeqRemove(heading, class_index);
+        }
+        else
+        {
+            if (!IsDefinedClass(ctx, SeqAt(list, class_index)))
+            {
+                SeqDestroy(list);
+                continue;
+            }
+
+            SeqRemove(list, class_index);
+            JsonElement *json_object = JsonObjectCreate(num_columns);
+
+            size_t const num_fields = SeqLength(list);
+            for (size_t i = 0; i < num_fields; i++)
+            {
+                if (has_heading)
+                {
+                    JsonObjectAppendString(json_object,
+                                           SeqAt(heading, i),
+                                           SeqAt(list, i));
+                }
+                else
+                {
+                    size_t const key_len = PRINTSIZE(size_t);
+                    char key[key_len];
+                    xsnprintf(key, key_len, "%zu", i);
+
+                    JsonObjectAppendString(json_object,
+                                           key,
+                                           SeqAt(list, i));
+                }
+            }
+
+            JsonArrayAppendObject(json, json_object);
+            SeqDestroy(list);
+        }
+    }
+
+    if (sort_arg != NULL)
+    {
+        size_t sort_index = IntFromString(RlistScalarValue(sort_arg));
+        if (sort_index == class_index)
+        {
+            Log(LOG_LEVEL_VERBOSE,
+                    "%s: sorting column (%zu) is the same as class "
+                    "expression column (%zu). Not sorting data container.",
```

Comment:
Why this limitation when `sort_arg` is optional anyway?

Reply:
IMO, if you want to keep this, it should be bumped up to WARNING log level. The user is specifying a sorting column which you have defined to be invalid.